File: samples/python/ws_contract_status.py
import os
import json
import signal
import sys
from datetime import datetime, timezone
from websocket import WebSocketApp
from dotenv import load_dotenv

load_dotenv()

GEMINI_WS_URL = os.getenv('GEMINI_WS_URL', 'wss://ws.gemini.com')


def stream_contract_status():
    """Stream prediction-market contract lifecycle events.

    Emits status transitions (e.g. Awaiting Approval -> Approved -> Active)
    and strike-populated moments for Up/Down contracts. The `p` field
    (strike price) is parsed from strike-based tickers; for Up/Down
    contracts it is omitted until the strike is set.
    """

    def on_open(ws):
        print(f'Connected to {GEMINI_WS_URL}')
        print('Subscribing to contractStatus...\n')

        ws.send(json.dumps({
            'id': '1',
            'method': 'subscribe',
            'params': ['contractStatus'],
        }))

    def on_message(ws, data):
        msg = json.loads(data)

        # Skip subscription confirmations
        if 'result' in msg or 'id' in msg:
            return

        time = datetime.fromtimestamp(msg['E'] / 1000, tz=timezone.utc).isoformat()
        strike = f" strike={msg['p']}" if msg.get('p') else ''
        print(f"[{time}] {msg['s']} [{msg['c']}] {msg['o']} -> {msg['n']}{strike}")

    def on_error(ws, err):
        print(f'WebSocket error: {err}')

    def on_close(ws, status_code, msg):
        print('Connection closed')

    ws = WebSocketApp(
        GEMINI_WS_URL,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close,
    )

    # Graceful shutdown
    signal.signal(signal.SIGINT, lambda *_: (print('\nClosing connection...'), ws.close(), sys.exit(0)))

    ws.run_forever()


if __name__ == '__main__':
    # Show help message
    if '--help' in sys.argv or '-h' in sys.argv:
        print('Usage: python3 ws_contract_status.py')
        print('Streams prediction-market contract lifecycle events (status transitions and strike-populated moments).')
        sys.exit(0)

    stream_contract_status()

File: samples/typescript/src/wsContractStatus.ts
import WebSocket from 'ws';
import * as dotenv from 'dotenv';

dotenv.config();

const GEMINI_WS_URL = process.env.GEMINI_WS_URL || 'wss://ws.gemini.com';

function streamContractStatus(): void {
  const ws = new WebSocket(GEMINI_WS_URL);

  ws.on('open', () => {
    console.log(`Connected to ${GEMINI_WS_URL}`);
    console.log('Subscribing to contractStatus...\n');

    ws.send(JSON.stringify({
      id: '1',
      method: 'subscribe',
      params: ['contractStatus'],
    }));
  });

  ws.on('message', (data: WebSocket.Data) => {
    const msg = JSON.parse(data.toString());

    // Skip subscription confirmations
    if (msg.result || msg.id) return;

    const time = new Date(msg.E).toISOString();
    // `p` (strike price) is parsed from strike-based tickers (e.g. HI78999D63).
    // Up/Down contracts publish `p` only once the strike is set.
    const strike = msg.p ? ` strike=${msg.p}` : '';
    console.log(`[${time}] ${msg.s} [${msg.c}] ${msg.o} -> ${msg.n}${strike}`);
  });

  ws.on('error', (err: Error) => {
    console.error('WebSocket error:', err.message);
  });

  ws.on('close', () => {
    console.log('Connection closed');
  });

  // Graceful shutdown
  process.on('SIGINT', () => {
    console.log('\nClosing connection...');
    ws.close();
    process.exit(0);
  });
}

if (process.argv.includes('--help') || process.argv.includes('-h')) {
  console.log('Usage: npx ts-node src/wsContractStatus.ts');
  console.log('Streams prediction-market contract lifecycle events (status transitions and strike-populated moments).');
  process.exit(0);
}

streamContractStatus();
